### DISCUSSION TITLE
Pelican Sprint

### DISCUSSION THREAD

**Discussion by justinmayer** [getpelican/pelican#3212] [Polls]
Who is up for a Pelican sprint? 👋 I think it would be great to get some folks together and collaborate on topics such as:

- Release Pelican 4.9
- Add support for Pip-installable themes (perhaps as [namespace packages](https://packaging.python.org/en/latest/guides/creating-and-discovering-plugins/#using-namespace-packages))
- Choose a theme to be first test case for above support; publish it to new [Pelican Themes](https://github.com/pelican-themes) organization
- Migrate more plugins to individual repos under the [Pelican Plugins](https://github.com/pelican-plugins) organization
- Maybe switch build/dev system to Hatchling + PDM/Hatch? ([context](https://github.com/getpelican/pelican/issues/3203#issuecomment-1765889516))
- Update docs re: how to develop Pelican, plugins, and themes
- Replace Flake8 & isort with Ruff and apply to entire code base
- Review and merge pull requests
- Review and close issues

What do you say about this coming weekend, October 28–29? Are you interested in lending a hand?

If so… **Would you please review the poll choices and choose the day+time slot during which you are most interested in participating?**

The poll options are shown in two different time zones: Western Europe (CET) and USA Pacific (PST)

Even if you are available and willing to participate for multiple day+slots, please choose the one that's best for you. Feel free to use this discussion thread to add any other thoughts and ideas about sprinting on Pelican.

Project to manage things: https://github.com/orgs/getpelican/projects/1

Milestones: [Pelican 4.9](https://github.com/getpelican/pelican/milestone/15) | [Pelican 5.0](https://github.com/getpelican/pelican/milestone/16)

This will be fun — I’m looking forward to it! 😁 

(I'm tagging some folks who might be interested — if not, I'm sorry for the noise! @getpelican/reviewers @kernc @kdeldycke @willthong @rlaboiss @maphew @galaxy4public @lgiordani @datatalking @seanh @silverhook @nandac @daturkel @offbyone @lioman @GiovanH @FriedrichFroebel @damascene @kzisme @mart-e @davidlesieur @Ceralor @arulrajnet)

**Comment by kernc** [comment]
I've been coming around for two years to even briefly describe my successful experiments with multilingual Pelican sites (without i18n-subsites or similar plugins). It's to expect my sprint contrib to be comparatively light. 😅

Are there any particular milestone issues/tags to be aware of?

Fwiw, I tend to oppose the new shiny gadgetry (hatch, ruff) until it's mature. A year ago I tried ruff: Two dependabot bumps per week for a few weeks EACH dragged in an extra handful of policy (code style) changes. 🤨 It's good to push the field forward, but I prefer not to do it at my expense. Hopefully you did the cost-benefit analysis. I'm still looking for a case that setup.py + wheel don't cover.

**Reply by justinmayer** [reply]
> Are there any particular milestone issues/tags to be aware of?

Excellent question! Thank you for finally giving me an excuse to try using GitHub's "Projects" feature. I just created one to coordinate things for the sprint: https://github.com/orgs/getpelican/projects/1

I also set up two milestones: [Pelican 4.9](https://github.com/getpelican/pelican/milestone/15) | [Pelican 5.0](https://github.com/getpelican/pelican/milestone/16)

One of the most helpful areas would be reviewing open pull requests, particularly [those that are already marked as "awaiting review"](https://github.com/getpelican/pelican/issues?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc+label%3A%22awaiting+review%22).

Another area that would be helpful is addressing [issues marked as related to documentation](https://github.com/getpelican/pelican/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Adocs).

Last but not least, helping to _identify_ which issues/PRs to add to the project/milestones would also be really useful. 😅 

> I tend to oppose the new shiny gadgetry (hatch, ruff) until it's mature

Your stance is understandable and full of wisdom. [Given that I myself have written about not being an early adopter](https://justinmayer.com/posts/early-adopter-tax/), you'd think I'd learn. You'd be wrong.

I jumped on the Poetry train early, and while in many ways I've been satisfied with it, I think it's time to move on. I personally prefer `pyproject.toml` to `setup.py`, but it's not something I feel very strongly about. I feel even less strongly about using Hatchling or Setuptools as the build system. More context here: https://github.com/getpelican/pelican/issues/3203#issuecomment-1765889516

To be fair to Ruff, I think the project was clear from the start that it was going to be a bumpy ride until a certain level of maturity arrived, and I think the formal version policy introduced as of the recently-released [Ruff 0.1](https://astral.sh/blog/ruff-v0.1.0) version is a step in the right direction.

**Comment by avaris** [comment]
My availability will be intermittent but I'll try to make time for PR review and issues. It's been on my list for a while anyway and it's starting to make me feel bad for delaying them due to not being able to make time 😓.

**Comment by offbyone** [comment]
I've been doing pdm/hatch conversions in lots of places, and am more than happy to tilt at that for Pelican.

**Reply by justinmayer** [reply]
Awesome! Maybe you and @lioman could work on that together. Have a look at the discussion in #3203 as well as the Hatchling + PDM experiments I've done in the Pelican Plugins org — for example, the [Featured Image plugin](https://github.com/pelican-plugins/featured-image).

**Reply by lioman** [reply]
I prepared already a branch. We can work on that together

**Comment by iKevinY** [comment]
I'm in a similar boat to @avaris. I really appreciate all the work that @justinmayer has been doing to continue to maintain Pelican after all these years (my blog is still powered by it!), but I haven't been involved in the project due to time constraints. If PRs or other comments float by as a result of this sprint, I'll do my best to chime in and give feedback as well, though!

**Comment by galaxy4public** [comment]
All,

I will try my best to carve out some time to help, but I got this
notification only today and it is a very short notice for me, since I
usually have the week planned in advance.

Also, I am located in Australia (haven't checked the proposed timeslots
yet, but usually it does not work very well with GMT+10).

**Comment by pauloxnet** [comment]
I like the idea of participating in a sprint on Pelican, unfortunately, I won't be able to join this weekend because I have to attend a conference. In general, however, it may be easier to participate during the working week. I'll follow developments asynchronously next week.

**Comment by maphew** [comment]
I applaud the event! Unsure if I will be free to join, likely not. I will however be cheering 📣

**Comment by mart-e** [comment]
I am not available this weekend unfortunately so won't be able to join but will see if I can check some PRs.
Thanks for the suggestion though.

> One of the most helpful areas would be reviewing open pull requests, particularly [those that are already marked as "awaiting review"](https://github.com/getpelican/pelican/issues?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc+label%3A%22awaiting+review%22).

Are the PRs with "awaiting review" waiting for technical reviews or also on the intent? If it's a new feature (e.g. #2761) does adding the tag mean you validate the idea and we can just look at the code or should we assume it's still not sure if this is something that you want in Pelican?

**Reply by justinmayer** [reply]
Excellent question, Martin. In short, I would answer your question with "both". Speaking for myself, for the most part I think you can assume that I'm open to merging the pull requests in the queue. In other words, I don't think there are any PRs there to which I am flatly opposed. At the same time, I would value the opinions of folks who provide feedback regarding the _added value_ of the PR and not just its technical merits. Does that make sense?

**Comment by lioman** [comment]
Voted for Sunday morning, but would be only potentially available in the first half of that slot

**Comment by offbyone** [comment]
Two things:

1. How is this sprint being done communication-tech-wise?
2. Am I supposed to be able to access the project link? I can't, if so.

**Reply by justinmayer** [reply]
Hey Chris! I was just about to update some info about that.

1. For now, let's use the IRC web-chat. Go to https://web.libera.chat/?#pelican and log into the `#pelican` channel.
2. Oops. I had no idea. Looking into that now. 😅

**Reply by justinmayer** [reply]
I fixed the project link, which should work now. Apparently they are private by default, but I've made it public. It's my first time using that feature, so it seems there is some learning to be done. 😊

**Comment by justinmayer** [comment]
Hello everyone! I am working on items in the [Pelican Sprint Project](https://github.com/orgs/getpelican/projects/1/views/1), which is by no means comprehensive, so if you want to suggest items that should be added, please do so! There is also the list of [open pull requests](https://github.com/getpelican/pelican/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc) and [open issues](https://github.com/getpelican/pelican/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc).

For real-time communication and coordination during the sprint, please come to the IRC web-chat at https://web.libera.chat/?#pelican and join the `#pelican` channel.

I will be spending as much time this weekend as I can on the sprint, dropping in and out when time affords. Looking forward to collaborating with you all! ✨

**Comment by nandac** [comment]
I have a minor request is it possible to get the `WRITE_SELECTED` feature working in the new release? 

It has been documented for years but does not work as per these open issues:
- https://github.com/getpelican/pelican/issues/2678

Just a suggestion for consideration. We could just remove it from the docs for now and add it back when we have it ready.

**Reply by justinmayer** [reply]
@nandac: Would you be willing to take a look and see if you can fix that feature?

**Reply by nandac** [reply]
@justinmayer I am willing to take a look at it however, I have a few questions.

The feature was documented to reduce the time taken to render pages in Pelican during development.

From the issue above it does not look like the patches that were tested did not really reduce the time taken by very much.

So I would like to know where Pelican spends its time and perhaps address that issue as well if possible.

**Reply by justinmayer** [reply]
[As suggested in that thread](https://github.com/getpelican/pelican/issues/2678#issuecomment-776614289), it might be better to approach this by switching focus from output path to source path. That seems to be a general point of agreement and seems to actually reduce generation time.

**Reply by mart-e** [reply]
I quickly checked a speedscope profile and confirms most of the time is spent in `generate_context` for me, not the `write_file`. I also tend to agree it would be more efficient to focus on limiting the generation of content instead of the output dir.

**Reply by avaris** [reply]
The problem with limiting read (a.k.a. `generate_context`) is that it's not trivial to figure out what is needed to generate a particular output file. Lets say you wrote a new article and want to generate only the output for that. Well... Your theme might have a sidebar to link to 5 other articles (newest, from a certain category, whatever). That means you also need to read those as well. So, in other words, there needs to be way for theme to communicate what resources are required to render a specific template.

**Comment by pauloxnet** [comment]
I found only now the time to propose an update of the "simple" theme that would be great to include in the Pelican 4.9 release.

See issue #3233 and PR #3234

**Comment by arulrajnet** [comment]
Hi @justinmayer I missed the sprint planning meeting. 

I would like to propose a new feature for the Pelican project. Currently, Pelican supports various objects, such as articles, authors, categories, and tags. However, at the moment, only articles have the option to include extra metadata through front-matter. I suggest that we enhance Pelican by allowing the addition of extra metadata for other objects through a configuration file. For instance, for authors, this metadata could include fields such as Full Name, Profile Image, and Short Bio.

**Reply by justinmayer** [reply]
Hi @arulrajnet. I think what I'm hearing is a request for a standardized setting that themes can rely upon to be consistent, similar to those that have historically existed: https://docs.getpelican.com/en/latest/settings.html#themes

The problem is that for something like, say, author-related metadata, it's hard to standardize. For example, last year a setting like `AUTHOR_TWITTER_USERNAME` made sense. This year... not so much.

Someone might want Threads, another BlueSky... It's difficult to standardize these things.

If anything, I think the preference is to _reduce_ this kind of thing — not increase it. At some point I am almost certainly going to remove [`DISQUS_SITENAME`](https://docs.getpelican.com/en/latest/settings.html#DISQUS_SITENAME) and [`TWITTER_USERNAME`](https://docs.getpelican.com/en/latest/settings.html#TWITTER_USERNAME), for example.

A bit of searching yielded related items you should look at:
 
- https://github.com/getpelican/pelican-plugins/tree/master/autopages
- https://github.com/pelican-plugins/theme-config (also discussed here: https://github.com/getpelican/pelican/pull/2719)
- https://github.com/getpelican/pelican/issues/2679

If you'd like to discuss this further, please open a separate Discussion topic, as discussing here in this thread will get unwieldy. 😊

**Comment by willthong** [comment]
Sorry to have missed this @justinmayer - had my notifications turned off. Keen for the next one!

**Reply by justinmayer** [reply]
No worries! I'd like to organize another sprint soon.

**Comment by GiovanH** [comment]
Also disappointed to have missed this. I've been very sick. I'll review some of the flagged pull requests when I get some time though! Appreciate you including me in the list

**Reply by justinmayer** [reply]
That's great. The help would be much appreciated. Feel better soon!

**Comment by pauloxnet** [comment]
I could be available next week to sprint a bit on Pelican to prepare a new release and/or work on Pelican 5 ( is this one the next release or we'll have 4.12?)

**Reply by justinmayer** [reply]
I am not yet sure whether the next release will be 4.12 or 5.0, but either way I think a sprint would be super helpful. I really appreciate your offer to assist! 😁

**Reply by pauloxnet** [reply]
> I am not yet sure whether the next release will be 4.12 or 5.0, but either way I think a sprint would be super helpful. I really appreciate your offer to assist! 😁 

I would suggest, a 4.12 with what we already have  in git. And after that unlock bigger feature (e.g. html5, ...) for 5.0
I can reserve some time to work on that next week. Unfortunately after that the conference season start 😅
